### ISSUE TITLE
Rate limit download speed on pulling new models

### ISSUE THREAD

**Issue by donuts-are-good** [jmorganca/ollama#2006]
Is there interest in implementing a rate limiter in the `pull` command? I'm open to working on this, this is the syntax I have in mind for now:

`ollama pull modelname --someflagname 1024` <-- this would limit to 1024 kbps

I took a look at the code in server/download.go, and I think I can do this with the x/time/rate applied to the downloadChunk method of the blob downloader.

This feature, or something like this that accomplishes this would be quite useful for me. Ollama is able to saturate my network faster than bit torrent or anything else I've tried.

**Comment by tkafka** [comment]
Yes, definitely! Same here - when I download models, everyone in the office gets a really slow internet.

How about `rate-limit`?

**Comment by jukofyork** [comment]
Yeah, same here!

I'm finding `ollama pull` is *really* killing my connection and i have to limit myself to just using it at night now...

I assume it's using multiple threads to download multiple chunks at the same time or something, as it seems a lot more lag-inducing than either `wget` or `curl`?  If so, then it might be good to have control over these parameter(s) too.

**Comment by donuts-are-good** [comment]
I'm still ready to implement. Holding for an OK from a maintainer before I create the PR.

**Comment by escaroda** [comment]
I would do the same as `wget`:

> ‘--limit-rate=amount’
>
> Limit the download speed to amount bytes per second. Amount may be expressed in bytes, kilobytes with the ‘k’ suffix, or megabytes with the ‘m’ suffix. For example, ‘--limit-rate=20k’ will limit the retrieval rate to 20KB/s. This is useful when, for whatever reason, you don’t want Wget to consume the entire available bandwidth.
> 
> This option allows the use of decimal numbers, usually in conjunction with power suffixes; for example, ‘--limit-rate=2.5k’ is a legal value.
> 
> Note that Wget implements the limiting by sleeping the appropriate amount of time after a network read that took less time than specified by the rate. Eventually this strategy causes the TCP transfer to slow down to approximately the specified rate. However, it may take some time for this balance to be achieved, so don’t be surprised if limiting the rate doesn’t work well with very small files.

**Comment by easp** [comment]
I think this is coming. I saw either a branch or a pull request to provide rate limiting by one of the maintainers.

**Comment by akulbe** [comment]
I would **_LOVE_** to see this implemented. It reliably and repeatedly kills my connection, on anything larger than a 13b model. I think it's the sustained speed (I have a 1G/1G connection, and downloads get up to 115M/s) when it happens.

**Comment by BruceMacD** [comment]
Behavior here will be improved by #2221, working on getting that unblocked now

**Comment by donuts-are-good** [comment]
We want to define an arbitrary download speed limit. It'd be great if #2221 could address that somehow.

**Comment by pablo-01** [comment]
Had this issue every day since starting using Ollama few days ago. In my case - Pop!_OS 22.04 LTS - freezes randomly and eventually freezes completely leaving no choice but to hard reset.

**Comment by fermuch** [comment]
@simmonsm I think trickle wouldn't work anyways since go doesn't use libc (since trickle uses LD_PRELOAD).

**Comment by simmonsm** [comment]
> @simmonsm I think trickle wouldn't work anyways since go doesn't use libc (and trickle uses LD_PRELOAD for its magic).

Fair enough. In the meantime I'm using a VM connected via a virtual traffic shaping network switch.

**Comment by LagSlug** [comment]
You might be able to accomplish this with a docker container

https://stackoverflow.com/questions/25497523/how-can-i-rate-limit-network-traffic-on-a-docker-container

**Comment by supercurio** [comment]
I'm downloading a bunch of Llama 3 models at the moment and last night my upstairs neighbor, which whom I'm sharing a 300/100 fiber connection asked for help because he couldn't use the internet anymore. Indeed, I ran a speedtest on another machine connected over Ethernet and the bandwidth left was 1.6Mbit/s download, with whopping ~1000ms of ping latency.

My Ollama instance is running on MacOS as a native app.
For now I found a workaround for my neighbor, using XCode's Network Link Conditioner, but I'm still essentially unable to browse the web on my primary machine when pulling models.

I appreciate that Ollama maximizes the bandwidth to download large models as quickly as possible, but the default behavior does not run with sane parameters at all.

My suggestion as a simple solution that can be implemented quickly:
- run by default with conservative settings to be a good network citizen. 2, maybe 3. Not more than that.
- mention a more aggressive preset as command-line argument when downloading via `ollama pull` or `ollama run`
- expose a custom "max concurrent download connections" parameter on command line and API.

Then later, sure - an adaptive algorithm can try to optimize the concurrent connection count based on latency and throughput. But it might never work that great on shared and mobile connections, where the available bandwidth and latency vary based on external parameters.
This github issue suggests a rate limit which would be helpful as well, but selecting an appropriate amount of concurrent connections should do the trick just fine without resorting to manual tuning.
If Ollama is competing with something else trying to use bandwidth, like a neighbor trying to watch Netflix, it should not try to hack TCP's congestion control standards to get all the bandwidth and respect them instead.

I hope this can be addressed shortly.

**Comment by mcraveiro** [comment]
Been hit by this issue when downloading large models, literally hogs the entire device. Would be nice to be able to limit the rate in someway a-la bittorrent clients.

**Comment by strangehelix** [comment]
Same issue. Downloader uses up the entire bandwidth (the internet becomes unusable) and eventually crashes because of the timeouts. Rate-limiter seems like a critically important feature.

**Comment by FeyNyXx** [comment]
Same issue. I'm killing my (not exactly mine, but I have to deal with it...) router using ollama pull :<

**Comment by metamec** [comment]
Love Ollama but this is murdering the end user experience for me.  I'm having to ctrl+c just to post this comment.  I have a 2Gb/s connection too.  It's not a limited bandwidth issue.  It's simply downloading too many chunks simultaneously, deprioritising internet bandwidth to every other process on the system.

**Comment by mcraveiro** [comment]
Yes, same here, I can only download models at night. Machine is unusable.

**Comment by MihailCosmin** [comment]
Had same problem, at work, not only my computer was getting slow but also all the internet for all other colleagues.
I had to find a solution to rate limit download speed, older solutions (wondershaper, trickle or tc) did not work for me. The only one that worked was [FireQOS](https://firehol.org/tutorial/fireqos-new-user/), in case anyone else needs it.

**Comment by LutzFassl** [comment]
+1

**Comment by robins** [comment]
My linux box (i5) got reliably stuck *every* *single* *time* I pulled a model... so +1 for the `--rate-limit` feature.

Two solutions, that did help me limp for now:
1. As soon as I started the fetch, I used `iotop` to change the `ionice` priority (using `i`) to `idle`. That made the issue completely go away in that, although the downloads were still fast, the linux system was quite usable. However, this was still frustrating since one had to type the PIDs when trying to set `ionice` for them (and there were a few)!
  
2. Now since `Ollama` spun up multiple downloads, the `ionice` tool didn't work for me - IIUC that's because `ionice` needs to be run for each process. So it ended up being far simpler to just get the parent PID, and then set ionice for each of the child processes, each time I was downloading a model. 

```
pid=`ps -ef | grep "ollama run" | awk '{print $2}'`
sudo ionice -c3 -p `ps -T -p $pid | awk '{print $2}' | grep -v SPID | tr '\r\n' ' '`
```

**Comment by treibholz** [comment]
I use this horrible "workaround", to not consume the whole internet bandwidth, so I can still work on my other machine, while pulling a model: 
```
$ sudo ethtool -s eth0  autoneg on speed 10 duplex full
```
This negotiates the linkspeed of my network-interface to 10Mbit.

Yes, the pulling machine itself is not useable as well and sometimes the download is interrupted, because (and I'm not kidding you!) there is not enough bandwidth left for DNS. But at least others on my local network are not angry anymore.

**Comment by supercurio** [comment]
I'm preparing a patch and will submit a PR to address this soon.

**Comment by Netzvamp** [comment]
My solution for now, works fine. This docker-tc can also simulate package loss 😂 

```yaml
version: '3'
services:
  ollama:
    image: ollama/ollama
    container_name: ollama
    ports:
      - 11434:11434
    restart: unless-stopped
    labels:
      - "com.docker-tc.enabled=1"
      - "com.docker-tc.limit=30mbit"

  docker-tc:
    image: lukaszlach/docker-tc
    cap_add:
      - NET_ADMIN
    network_mode: host
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /var/docker-tc:/var/docker-tc
```

**Comment by supercurio** [comment]
To everyone in this thread, I'd encourage you to build your own Ollama from my branch for testing purposes and report how the issue is solved. I'm curious how much of your available bandwidth is used during downloads with the new default.

For me, on MacOS, making a build was easy following https://github.com/ollama/ollama/blob/main/docs/development.md
I see that many PRs are awaiting review and merging, so I don't know how long it'll take.
However, using Ollama is so annoying until this issue is solved, I'm determined to make this fix happen.

**Comment by treibholz** [comment]
@supercurio Works great here. I can still download with 11MB/sec on my 100mbit line here, the machine is still responsive AND I can still watch something in HD over a streaming-service.

**Comment by binarynoise** [comment]
I can confirm as well that reducing the amount of parallel connections restores network usability (even though I patched the source by hand and not using your feature branch).
Interestingly the download speed is unaffected (maxes at 20MB/s which I think is a server limit).

**Comment by scscgit** [comment]
I'm adding a vote, it's really disruptive when you can't even join a meeting in work due to slow internet; some users may not be able to figure out the root cause. I ran it in Docker; even after pausing container/engine, it still consumed the entire bandwidth, so I had to either wait some more or completely shut down the Docker. Note that our scenario could be using a tool like Open WebUI to download the models, so it's not enough to provide hidden CLI parameters, and we need a quick solution that we won't need to spend time googling; Ollama should properly display a warning if using it may cause such disruptions.

**Comment by enrico3** [comment]
I am using @Netzvamp 's solution. I added these parameters, maybe it helps someone:
 ```yaml
  ollama:
     [...]
     volumes:
     #with this bind the container downloads new models into my existing model directory on my host
      - /home/USERNAME/.ollama:/root/.ollama
     environment:
     #do not prune partly downloaded models when starting the container. 
     #This way downloads do not need to be completed in one session.
      - OLLAMA_NOPRUNE=true
 ```

The docker-tc container needs to be started before the ollama container I think.

Sometimes my download stopped because of a TLS handshake timeout. So I put the command in a loop to immediately resume it after an error:
```
#!/bin/bash
ollama pull MODELNAME
while [ $? -ne 0 ]; do
    ollama pull MODELNAME
done
```
With this command on the host the speed limit can be changed while the download is running (https://github.com/lukaszlach/docker-tc#post)
`curl -d'rate=20Mbit' localhost:4080/ollama`

**Comment by Fluffkin** [comment]
For those on Linux "Traffic Toll" https://github.com/cryzed/TrafficToll sort of works. But I gave up on ollama because even with that some segments get zero data for long enough to trigger timeouts, so even with a fully saturated bandwidth the download fails. :)

I'm puzzled by two things:
Why hasn't Huggingface politely asked to stop opening stupid amounts of connections? It's generally seen as bad internet etiquette to attempt to hog an unnecessary amount of connections for a download.
What's the reasoning / use case behind the way the download threads are currently handled in ollama? Is the dev using a near internet backbone speed connection where the bandwidth used doesn't effect anything or anybody else?

**Comment by joelanman** [comment]
just to note as I don't think its clearly in this thread - the issue isn't rate limiting per se - it downloads at 10mbps for me. It's that it is setting up 64 concurrent connections to do so, as per the PR here:

https://github.com/ollama/ollama/pull/5683

**Comment by igorschlum** [comment]
@joelanman I agree with you. From the same location, I was able to download a 3K model but not the 405b model. For the 3K model, the 3GB were downloaded without any connection issues, whereas with the 405b model, the download kept stopping after about every 200MB.

**Comment by ShayBox** [comment]
This reliably crashes my router and causes it to restart, it's too fast.

**Comment by numbermaniac** [comment]
> I'm having to ctrl+c just to post this comment.

Same here. I literally can't even search Google while it's downloading something. I can download files that are multiple gigabytes in my web browser or in macOS Homebrew and still use the internet just fine, but when Ollama is downloading a model, my entire internet becomes unusable.

It doesn't even work well for Ollama either because the download speed starts at around 6MB/s and then just keeps gradually reducing, going down to like 1MB/s, then 500 KB/s, then 350 KB/s, gradually getting slower and slower, so Ollama is practically sabotaging _itself_ with the way it downloads files.

**Comment by supercurio** [comment]
I wonder how to get some progress going on this issue: the description of how badly it affects the user experience should bump its priority to critical in my opinion, and I already submitted a PR with a fix a month ago.

After testing Ollama on a high bandwidth VPS, I appreciated that its aggressive strategy lead to 300MB/s (mega bytes) downloads, so I get the point on keeping the many concurrent connections capability available. Which is the case with my patch.

How to more forward from here?
Routers crashing, people unable to use their computers: that's not the expected result when using any kind of software.

**Comment by igorschlum** [comment]
Hi @supercurio (bonjour François), your patch is here: 
https://github.com/ollama/ollama/pull/5683
I know that there were issues that were more important during those weeks, like CUDA fixes, Memory and function calling.
Ollama can run easily without this patch, but if it's just a matter of approving your patch, @jmorganca could decide to do it.

**Comment by supercurio** [comment]
Salut @igorschlum 😌
All of Ollama's the core functionalities are important, that's for sure.
Downloading model(s) is still the first action every Ollama user will take.

I solved for my individual use case already but I'm hoping to use Ollama as LLM runtime for the app I'm developing at the moment. It's a non-starter until this issue is solved unfortunately.

**Comment by joelanman** [comment]
Is it fixed by this?

 - https://github.com/ollama/ollama/pull/6347

**Comment by Fluffkin** [comment]
It'll help in many cases, but it's probably still too high for some domestic broadband users with low bandwidth. ¯\_(ツ)_/¯

**Comment by igorschlum** [comment]
@Fluffkin I think it would work for any type of connection, because actually Ollama is downloading 64 files simultaneously and letting very little bandwidth for other computers. With this modification, it will use only one download at a time.

**Comment by joelanman** [comment]
@igorschlum no it's changed from 64 to 16, so still a lot of connections

**Comment by robins** [comment]
My 2c. Unless there a way for customers to "request" for more speed, the default shouldn't be hurting low-end users.

I'd take Git's example here. Albeit downloads are sequential, but even on a 20 core machine, it doesn't spin up an obscene number of processes - only 4 processes - to churn through the repo. 

So yes, we shouldn't cut down from 64 to 1, just to accommodate everyone - but settling for something like **default of 4 threads sounds like a good middle-ground here**.... If / when there's a feature to request for more threads, customers with more resources can always go back to 64 concurrent threads!

**Comment by igorschlum** [comment]
@mrtysn what version of Ollama are you using? I agree with @supercurio that a parameter could be added to set the number of concurrent downloads.
What version of Ollama do you use? On which OS?

**Comment by mrtysn** [comment]
> @mrtysn what version of Ollama are you using? I agree with @supercurio that a parameter could be added to set the number of concurrent downloads. What version of Ollama do you use? On which OS?

@igorschlum  Apologies for the lack of details.

- I am on a MacBook Pro M2 Max with Sonoma 14.6.1.
- My ollama is installed from homebrew, and it is currently on version 0.3.9.

However, I believe I did ~90% of my model pulls while on version 0.3.8, the homebrew cask was updated to 0.3.9 [~2 days ago](https://github.com/Homebrew/homebrew-core/commits/87cbc9cf2ccc2b83ae27821cb7203e3d416d71e1/Formula/o/ollama.rb).

<img width="438" alt="image" src="https://github.com/user-attachments/assets/56a334b4-210d-4f9a-99a9-c4db1b4aceac">

**Comment by igorschlum** [comment]
@mrtysn I installed Go on my Mac and was able to build Ollama from the source. If you'd like, I can create a tutorial on how to do this from scratch on a Mac. Then, you could change the 4 simultaneous downloads to 1 to see if this fixes the issue you're facing.

**Comment by joelanman** [comment]
@igorschlum it's 16, which is still very high

**Comment by igorschlum** [comment]
@joe Sorry, I anticipated that it could be 4 :-)

**Comment by mrtysn** [comment]
> @mrtysn I installed Go on my Mac and was able to build Ollama from the source. If you'd like, I can create a tutorial on how to do this from scratch on a Mac. Then, you could change the 4 simultaneous downloads to 1 to see if this fixes the issue you're facing.

I've installed the pre-requisites. However, unfortunately, I will not be able to attend more to this issue with my current workload. Scheduling model downloads for the night has been an okay workaround so far, and I currently have most of the models that I would like to utilize.

Nevertheless, I still think a CLI flag with reasonable defaults to control the number of concurrent download threads would be helpful for all ollama users, especially new users who would be downloading models for the first time or users from regions with slower internet speeds.

**Comment by donuts-are-good** [comment]
Hello all,

I appreciate everyone's input in this thread, and I do hope this eventually gets solved. When I initially created this issue as a place to discuss and outline a solution to the network saturation issue, I assumed it would be welcome with open arms. In the time since, the software has got bigger and more complex and it's no longer a simple solution, and doesn't appear to be a priority.

With that in mind, life must go on. I'm withdrawing the offer to implement this as I no longer have the resources to write and test a feature that isn't a concern of the developers. I don't think anybody's going to worry about it, but I wanted to be clear in my intentions. I look forward to seeing this issue fixed some day.

**Comment by ShayBox** [comment]
The issue was fixed 2 weeks ago in 0.3.7...

**Comment by mdlmarkham** [comment]
I'm still having the issue.

**Comment by devrandom** [comment]
it would be best if the number of threads was configurable.  most of these issues can be mitigated by setting the number of concurrent downloads to 1.

**Comment by jimbothegrey** [comment]
used wondershare to slow down the connection. looks like it is working....

**Comment by mrtysn** [comment]
> used wondershare to slow down the connection. looks like it is working....

@jimbothegrey what might wondershare be? all I'm finding is a file converter

**Comment by TiddlyWiddly** [comment]
Still experiencing this on pretty quick connection, knocks all my devices off

**Comment by ankh2054** [comment]
yeah same here, It's hard to work on anything else when models download and very large models will take multiple days to download at my 3.5MB/s

**Comment by xtareq** [comment]
Same issue here. But in my case it was good in my windows 11 for llama3.1 but the issue arise when i try to pull phi4. Any specific reason why this happens?

**Comment by ading2210** [comment]
As a rudimentary workaround, I wrote a bash script that monitors ollama's network usage and constantly suspends/resumes the ollama process.

https://gist.github.com/ading2210/882565526f7e1f2b9b14a022ac3741ac

Make sure you have `nethogs` installed (`sudo apt install nethogs`).

For example, if you want to limit downloads to 5000 KB/s:
```
$ sudo ./ollama-limiter.sh 5000
```

**Comment by martinoturrina** [comment]
bump

**Comment by martinoturrina** [comment]
bump

**Comment by joelanman** [comment]
that wouldnt fix it for people not using docker?

**Comment by pleoset** [comment]
I've tried limiting ollama.exe on Windows 11 Home via PowerShell with QoS Policy, and when pulling a model (phi4), in the Task Manager shows my new cap for the ollama.exe, but in fact in the CLI you can see that Ollama is blasting full bandwidth download and everything at home is starving. for bandwidth. Not to mention that I have Router level QoS policy implemented, but Ollama does not seem to care at all What is going on here? :D

**Comment by sirnails** [comment]
Bump

**Comment by whjvenyl** [comment]
Mac version of a rate limiter based on the previous script from @ading2210 

```
#!/bin/bash
#
# ollama-limiter.sh (v6 - Rate Calculation Method)
#
# This version calculates bandwidth as a rate (KB/s) between two snapshots,
# which is the most accurate method. It correctly parses the comma-delimited
# output from `nettop -L 1` and uses `tput` to prevent display issues.
#

# --- Configuration ---
MAX_BANDWIDTH_KBPS=8000
CHECK_INTERVAL=2 # Seconds between measurements
SUSPEND_TIME=4
RESUME_TIME=1

# --- Colors ---
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[1;33m'
BLUE='\033[0;34m'
NC='\033[0m'

# --- Functions ---

# This function runs nettop and extracts the PID and total cumulative bytes
# for the ollama process with the most network activity.
get_ollama_stats() {
    # -L 1: Run for one snapshot.
    # -J bytes_in,bytes_out: Output only these columns for easier parsing.
    nettop -L 1 -J bytes_in,bytes_out | grep 'ollama\.' | awk -F, '
    {
        # $1 is process.pid, $2 is bytes_in, $3 is bytes_out
        split($1, p, ".");
        pid = p[2];
        gsub(/[^0-9]/, "", pid); # Sanitize PID to ensure it is only numeric

        total_bytes = $2 + $3;

        # Print total bytes and pid, so we can sort to find the one with max traffic
        printf "%d %s\n", total_bytes, pid;
    }' | sort -rn | head -n 1
}

# --- Main ---

echo -e "${GREEN}Ollama Bandwidth Limiter is running.${NC}"
echo -e "Limit set to: ${YELLOW}${MAX_BANDWIDTH_KBPS} KB/s${NC}"
echo "Press Ctrl+C to stop."
echo ""

# Initialize stats for rate calculation
last_bytes=0
last_pid=""
last_time=$(date +%s.%N)

while true; do
    # Use tput to prevent terminal display issues.
    # `cr` = carriage return, `el` = erase line.
    tput cr
    tput el

    # Get current stats
    current_stats=$(get_ollama_stats)
    current_time=$(date +%s.%N)

    if [ -z "$current_stats" ]; then
        echo -n -e "${BLUE}[$(date '+%H:%M:%S')] Waiting for an active 'ollama' process...${NC}"
        sleep "$CHECK_INTERVAL"
        continue
    fi

    current_bytes=$(echo "$current_stats" | awk '{print $1}')
    current_pid=$(echo "$current_stats" | awk '{print $2}')

    # If the PID changed (e.g., new pull command) or this is the first run,
    # reset the baseline to start a new measurement.
    if [ "$current_pid" != "$last_pid" ]; then
        echo -n -e "${YELLOW}[$(date '+%H:%M:%S')] New active PID: ${current_pid}. Initializing...${NC}"
        last_pid=$current_pid
        last_bytes=$current_bytes
        last_time=$current_time
        sleep "$CHECK_INTERVAL"
        continue
    fi

    # --- Calculate Bandwidth Rate ---
    time_diff=$(echo "$current_time - $last_time" | bc)
    byte_diff=$(echo "$current_bytes - $last_bytes" | bc)

    # Avoid division by zero if the time difference is too small
    if (( $(echo "$time_diff <= 0.001" | bc -l) )); then
        kbps=0
    else
        # Rate in Bytes/sec, then convert to KB/s
        bytes_per_second=$(echo "$byte_diff / $time_diff" | bc)
        kbps=$(echo "$bytes_per_second / 1024" | bc)
    fi

    # Ensure kbps is a non-decimal integer
    kbps=${kbps%.*}
    if [ -z "$kbps" ]; then kbps=0; fi

    echo -n -e "${GREEN}[$(date '+%H:%M:%S')] Active PID: ${current_pid}, Bandwidth: ${kbps} KB/s${NC}"

    # --- Throttle if needed ---
    if (( kbps > MAX_BANDWIDTH_KBPS )); then
        echo "" # Newline for the suspend message
        echo -e "${RED}Limit exceeded! Suspending process ${current_pid} for ${SUSPEND_TIME}s...${NC}"
        kill -STOP "$current_pid"
        sleep "$SUSPEND_TIME"

        echo -e "${GREEN}Resuming process ${current_pid}...${NC}"
        kill -CONT "$current_pid"

        # Reset baseline after resuming to avoid a huge artificial spike on the next check
        last_pid=""
        sleep "$RESUME_TIME"
    else
        # Update baseline for the next iteration
        last_pid=$current_pid
        last_bytes=$current_bytes
        last_time=$current_time
        sleep "$CHECK_INTERVAL"
    fi
done

```

**Comment by codeisnotcode** [comment]
Rate-limiting is a really important feature! With the source code it should be simple to implement and it would provide high value. It is super annoying to have Ollama kill other network activity every time I download a new model.

**Comment by sotander** [comment]
I have this issue also. I pull at ca. 200MB/s. The admins are sending me emails to not slow down the network.

**Comment by Mondonno** [comment]
I have this issue also, it hurts especially on slower networks in my case, seems related also to [#3741](https://github.com/ollama/ollama/issues/3741)

**Comment by anton-karlovskiy** [comment]
@jmorganca 

👋 I’ve been trying to download llama3.2 on Windows 11 (ollama run llama3.2) and keep running into issues:

- An existing connection was forcibly closed by the remote host
- Error: max retries exceeded

What I’ve tried so far:

- https://github.com/ollama/ollama/issues/3769#issuecomment-2076767384: closed VPN → no change
- https://github.com/ollama/ollama/issues/6211#issuecomment-2626567037: changed DNS → made some progress (got to ~37%) but stuck again

<img width="1086" height="124" alt="Image" src="https://github.com/user-attachments/assets/774acce1-b0af-4b66-b2ae-415baf6f0fe0" />

From what I can tell, the main culprit seems to be an unstable connection. The workaround right now is to just re-run the command until it eventually completes, but this gets tedious.

To automate retries, I wrapped it in a small script. Sharing here in case it helps others or the devs want to consider integrating retry logic directly:

**Windows**
Open new notepad and put this code.

```bash
@echo off
:loop
echo Running ollama run llama3.2...
ollama run llama3.2

if %errorlevel% neq 0 (
    echo The command failed. Retrying... Press Ctrl+C to exit.
    goto loop
)
echo The command completed successfully!
```

Next go to save as… and pick All files as file type and name the file with `.bat` extension.
And finally you can run it.
To edit the code above, just change model name from `llama3.2` to any other model you like.

**Linux**
I didn’t test the code for Linux but this should hopefully work.

#!/bin/bash
```bash
# Function to handle Ctrl+C
trap 'echo "Script terminated by user. Exiting..."; exit' SIGINT
while true; do
    echo "Running ollama run llama3.2..."
    ollama run llama3.2
    if [ $? -eq 0 ]; then
        echo "The command completed successfully!"
        break
    else
        echo "The command failed. Retrying... Press Ctrl+C to exit."
    fi
done
```

Save this as `run_loop.sh` then run these commands.

```bash
chmod +x run_loop.sh
```
```bash
./run_loop.sh
```

Re: https://medium.com/@timnirmal/ollama-max-retries-exceeded-error-de6e0f86383e

Good luck 🤞

**Comment by ading2210** [comment]
@anton-karlovskiy On Linux there is a [`retry` command](https://manpages.debian.org/trixie/retry/retry.1.en.html) already, so there's no need for a shell script. 

You would just use it like:
```
retry ollama run llama3
```

**Comment by anton-karlovskiy** [comment]
> [@anton-karlovskiy](https://github.com/anton-karlovskiy) On Linux there is a [`retry` command](https://manpages.debian.org/trixie/retry/retry.1.en.html) already, so there's no need for a shell script.
> 
> You would just use it like:
> 
> ```
> retry ollama run llama3
> ```

Thank you for your tip. @ading2210 
Let's keep in touch. :)

**Comment by digitalextremist** [comment]
Agreed with @anton-karlovskiy, that was a real-life `Pro Tip` ™️ @ading2210 ... extra `rtfm` points there.

And nice work on the meticulousness, and DIY solutions @anton-karlovskiy. Do you ever use containers?

**Comment by PeaStew** [comment]
coming up to 2 years now and no proper solution, just hacks

**Comment by apassi** [comment]
I am using the trickle to limit the bandwith:
trickle -s -d 50mb ollama pull xxxxx

**Comment by floriandotorg** [comment]
Super important feature, not sure, why it's not here.

**Comment by dhirajlochib** [comment]
I've opened a PR to address this: #15219

The implementation adds a new `OLLAMA_MAX_DOWNLOAD_SPEED` environment variable that lets you cap download bandwidth when pulling models. It uses a shared token-bucket rate limiter (`golang.org/x/time/rate`) across all 16 concurrent download chunks, so the aggregate bandwidth stays within the specified limit.

Usage examples:
- `OLLAMA_MAX_DOWNLOAD_SPEED=10m` — limit to 10 MB/s
- `OLLAMA_MAX_DOWNLOAD_SPEED=500k` — limit to 500 KB/s
- `OLLAMA_MAX_DOWNLOAD_SPEED=1g` — limit to 1 GB/s

When unset, downloads run at full speed with zero overhead.

**Comment by donuts-are-good** [comment]
It's crazy that this issue is still open, still being replied to, and still no solution.

**Comment by phamngocduy98** [comment]
yes, I have the same issue when uploading also. My internet will fail with 16 concurrent connections.

**Comment by gregs007** [comment]
@dhirajlochib Thanks for developing a solution for this problem for us. It's been a nagging issue for quite some time. Really appreciate the contribution. Looking forward to it!

**Comment by Legendary-Lava** [comment]
A hack I did was isolating by source/server on an IFB (Intermediate function block) that is set below my max bandwidth.

Modifying anything on ingress is always a hack, but its a little more universal than addressing ollama spamming 16 connections specifically.
It is more prone to multiple different "server" connections like when torrenting
make to change eth2 to the correct interface & set bandwidth somewhere below your actual speed
```
ip link add name ifb4eth2 type ifb
tc qdisc del dev eth2 ingress
tc qdisc add dev eth2 handle ffff: ingress
tc qdisc del dev ifb4eth2 root
tc qdisc add dev ifb4eth2 root cake bandwidth 200Mbit dual-srchost besteffort egress
ip link set ifb4eth2 up # if you don't bring the device up your connection will lock up on the next step.
tc filter add dev eth2 parent ffff: matchall action mirred egress redirect dev ifb4eth2
```
To test just go to fast.com set it to 30 streams, extend test duration & try to do anything else thoughout the download.

**Comment by gregs007** [comment]
Maybe this will help someone until this change gets into main

I use a tool called tc to limit the ollama container after it starts. I'm no developer so I'm sure someone could make this more universal by automagically parsing out the right interface, burstsize, etc. but this limits my ollama network speed to 500m (I have 1g internet).

```
INTERFACE_ID=$(docker exec ollama cat /sys/class/net/eth0/iflink)
VETH=$(ip link show | grep "^${INTERFACE_ID}:" | sed -n 's/.*\(veth[a-z0-9]*\).*/\1/p')

tc qdisc add dev $VETH root tbf rate 500mbit burst 1mb latency 50ms
```

**Comment by Legendary-Lava** [comment]
I know this doesn't add much to discussion but too many concurrent connections should not be practiced. Just one or two connections would get a high enough bandwidth to be negligible outside of high loss environments.
Knocking out a network for 1 hour is far worse than slowing it down for an hour 15 minutes. Opportunity cost is an important factor.

My case is particularly egregious at 14% retransmits according to wireshark. I would strongly encourage measurements of retransmits; downloads of anything in general should be nowhere near that amount.

**Comment by Torchwood42** [comment]
I am also having this issue, very frustrating, the download will also sometimes revert, I had a model downloaded at around 90% and then it jumped to below 20% somehow. Almost every model download fails for the same reason in this issue.

**Comment by Kenqr** [comment]
Noooo don't fix it! I use ollama to cut off internet of my family and office, and no one can find evidence on my computer.  Who would expect ollama is the culprit, after all?
/s

**Comment by kabanowster** [comment]
the lack of env setting limiting the number of concurrent connections is dumb as hell
please make it happen
